session/direct-ipc.c

```diff
@@ -0,0 +1,241 @@
+#define _POSIX_C_SOURCE 200809L
+#include <errno.h>
+#include <stdio.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/socket.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/sysmacros.h>
+#include <sys/wait.h>
+#include <xf86drm.h>
+#include <linux/major.h>
+#include <wlr/util/log.h>
+#include "session/direct-ipc.h"
+
+enum { DRM_MAJOR = 226 };
+
+#ifdef HAS_LIBCAP
+#include <sys/capability.h>
+
+static bool have_permissions(void) {
+	cap_t cap = cap_get_proc();
+	cap_flag_value_t val;
+
+	if (!cap || cap_get_flag(cap, CAP_SYS_ADMIN, CAP_PERMITTED, &val) || val != CAP_SET) {
```

Comment:
Capabilities are just splitting up the special things root can do, so naturally, root has all of them. I would say try the examples as root, but it's currently it's broken at the moment and will probably require you to restart your computer. The breakage isn't part of this code though: it successfully gets past it, and also breaks on the logind backend.

Unfortunately it has to be CAP_SYS_ADMIN. The DRM code in linux explicitly checks for it, and if we don't handle being the DRM master properly, we will be unable to change VTs.